### PR TITLE
Feature/s3 session token

### PR DESCRIPTION
**Pull request by shanielh**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow passing optional SESSION_TOKEN to `s3` table function

### Documentation entry for user-facing changes

- [X] Documentation is written (mandatory for new features)
- [ ] I need help with ru documentation 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 324a578b50dfbb9f9a75d322c93165e0f536ed0b with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57850/324a578b50dfbb9f9a75d322c93165e0f536ed0b/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7206409116">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57850/324a578b50dfbb9f9a75d322c93165e0f536ed0b/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/IO/S3Common.cpp [hunk 1]
```diff
@@ -133,7 +135,7 @@ AuthSettings AuthSettings::loadFromConfig(const std::string & config_elem, const
 
     return AuthSettings
     {
-        std::move(access_key_id), std::move(secret_access_key),
+        std::move(access_key_id), std::move(secret_access_key), std::move(session_token),
```
Comment (Avogar) [inline src/IO/S3Common.cpp:138]: We should also set `session_token` below in the function `AuthSettings::updateFrom`

File: src/Storages/StorageS3.cpp [hunk 2]
```diff
@@ -1541,7 +1541,7 @@ StorageS3::Configuration StorageS3::getConfiguration(ASTs & engine_args, Context
         static std::unordered_map<size_t, std::unordered_map<std::string_view, size_t>> size_to_engine_args
         {
             {1, {{}}},
-            {5, {{"access_key_id", 1}, {"secret_access_key", 2}, {"format", 3}, {"compression_method", 4}}}
+            {6, {{"access_key_id", 1}, {"secret_access_key", 2}, {"session_token", 3}, {"format", 4}, {"compression_method", 5}}}
```
Comment (Avogar) [inline src/Storages/StorageS3.cpp:1544]: Please, update the list of supported signatures above in the comment

File: src/Storages/StorageS3.cpp [hunk 3]
```diff
@@ -1590,7 +1591,32 @@ StorageS3::Configuration StorageS3::getConfiguration(ASTs & engine_args, Context
...
+            {
+                auto fourth_arg = checkAndGetLiteralArgument<String>(engine_args[3], "session_token/format");
+                if (fourth_arg == "auto" || FormatFactory::instance().getAllFormats().contains(fourth_arg))
+                {
+                    engine_args_to_idx = {{"access_key_id", 1}, {"secret_access_key", 2}, {"session_token", 3}};
+                }
+                else
+                {
+                    engine_args_to_idx = {{"access_key_id", 1}, {"secret_access_key", 2}, {"format", 3}};
+                }
```
Comment (Avogar) [inline src/Storages/StorageS3.cpp:1603]: Seems like you mixed up `if/else` branches here, when `if` condition is true, the fourth argument should be `format`

File: src/Storages/StorageS3.cpp [hunk 4]
```diff
@@ -1590,7 +1591,32 @@ StorageS3::Configuration StorageS3::getConfiguration(ASTs & engine_args, Context
...
+        {
+            auto fourth_arg = checkAndGetLiteralArgument<String>(engine_args[3], "session_token/format");
+            if (fourth_arg == "auto" || FormatFactory::instance().getAllFormats().contains(fourth_arg))
+            {
+                engine_args_to_idx = {{"access_key_id", 1}, {"secret_access_key", 2}, {"session_token", 3}, {"format", 4}};
+            }
+            else
+            {
+                engine_args_to_idx = {{"access_key_id", 1}, {"secret_access_key", 2}, {"format", 4}, {"compression", 5}};
+            }
```
Comment (Avogar) [inline src/Storages/StorageS3.cpp:1619]: Again seems like `if/else` branches are mixed up

File: src/TableFunctions/TableFunctionS3.h [hunk 5]
```diff
@@ -26,7 +26,8 @@ class TableFunctionS3 : public ITableFunction
                                       " - url, format, structure, compression_method\n"
                                       " - url, access_key_id, secret_access_key, format\n"
                                       " - url, access_key_id, secret_access_key, format, structure\n"
-                                      " - url, access_key_id, secret_access_key, format, structure, compression_method\n"
+                                      " - url, access_key_id, secret_access_key, format, structure, compression_method\n",
+                                      " - url, access_key_id, secret_access_key, session_token, format, structure, compression_method\n"
```
Comment (Avogar) [inline src/TableFunctions/TableFunctionS3.h:30]: I think we should also support signature `url, access_key_id, secret_access_key, session_token, format`

File: src/TableFunctions/TableFunctionS3.cpp [hunk 6]
```diff
@@ -118,9 +118,10 @@ void TableFunctionS3::parseArgumentsImpl(ASTs & args, const ContextPtr & context
...
                 args_to_idx = {{"access_key_id", 1}, {"secret_access_key", 2}};
         }
-        /// For 4 arguments we support 3 possible variants:
+        /// For 4 arguments we support 4 possible variants:
         /// - s3(source, format, structure, compression_method),
-        /// - s3(source, access_key_id, access_key_id, format)
+        /// - s3(source, access_key_id, access_key_id, format),
+        /// - s3(source, access_key_id, access_key_id, session_token)
         /// - s3(source, NOSIGN, format, structure)
         /// We can distinguish them by looking at the 2-nd argument: check if it's a format name or not.
```
Comment (Avogar) [inline src/TableFunctions/TableFunctionS3.cpp:126]: Update this comment, now we also should check 4-th argument if 2-nd is not format

File: src/TableFunctions/TableFunctionS3.cpp [hunk 7]
```diff
@@ -132,12 +133,25 @@ void TableFunctionS3::parseArgumentsImpl(ASTs & args, const ContextPtr & context
...
+                    args_to_idx = {{"access_key_id", 1}, {"secret_access_key", 2}, {"session_token", 3}};
+                }
+            }
         }
-        /// For 5 arguments we support 2 possible variants:
+        /// For 5 arguments we support 3 possible variants:
         /// - s3(source, access_key_id, access_key_id, format, structure)
+        /// - s3(source, access_key_id, access_key_id, session_token, format)
         /// - s3(source, NOSIGN, format, structure, compression_method)
         /// We can distinguish them by looking at the 2-nd argument: check if it's a NOSIGN keyword name or not.
```
Comment (Avogar) [inline src/TableFunctions/TableFunctionS3.cpp:156]: Update this comment too

File: src/TableFunctions/TableFunctionS3.cpp [hunk 8]
```diff
@@ -149,7 +163,31 @@ void TableFunctionS3::parseArgumentsImpl(ASTs & args, const ContextPtr & context
...
+                }
+                else
+                {
+                    args_to_idx = {{"access_key_id", 1}, {"secret_access_key", 2}, {"session_token", 3}, {"format", 4}};
+                }
+            }
+        }
+        // For 6 arguments we support 2 possible variants:
+        /// - s3(source, access_key_id, access_key_id, format, structure, compression_method)
+        /// - s3(source, access_key_id, access_key_id, session_token, format, structure)
```
Comment (Avogar) [inline src/TableFunctions/TableFunctionS3.cpp:180]: Please, write a comment how we distinguish these 2 variants

**Comment by Avogar** [comment]
We should add a test for each new signatures with `session_token`, and a test with named collection containing `session_token`. Probably the best way is to add an integration test (see test[ test_storage_s3/test.py::test_signatures](https://github.com/ClickHouse/ClickHouse/blob/325374c68b5440eae60836b880bac42583fc53dc/tests/integration/test_storage_s3/test.py#L1397) as an example). Not sure if it will be easy to test the real session token (can we setup it in minio that is used in integration tests?).

**Comment by shanielh** [comment]
> We should add a test for each new signatures with `session_token`, and a test with named collection containing `session_token`. Probably the best way is to add an integration test (see test[ test_storage_s3/test.py::test_signatures](https://github.com/ClickHouse/ClickHouse/blob/325374c68b5440eae60836b880bac42583fc53dc/tests/integration/test_storage_s3/test.py#L1397) as an example). Not sure if it will be easy to test the real session token (can we setup it in minio that is used in integration tests?).

Hi, I think I did everything except the tests, I didn't manage to run them locally, is there a guide somewhere so I can try to test that locally, even testing that a faulty session_token is passed should be sufficient, right?

**Comment by Avogar** [comment]
> Is there a guide somewhere so I can try to test that locally

https://github.com/ClickHouse/ClickHouse/blob/master/tests/integration/README.md

> even testing that a faulty session_token is passed should be sufficient, right?

Yes, even if MinIO doesn't support session_token, we should at least check that all new signatures with session_token works

**Comment by Avogar** [comment]
> Any chance this works on Mac OS too?

Don't know, sorry. You can try, if you will have any problems, you can share them here, I will ask my colleagues

**Comment by shanielh** [comment]
> Don't know, sorry. You can try, if you will have any problems, you can share them here, I will ask my colleagues

I'm getting the following error: 

> Can't start docker daemon, timeout exceeded.

Note: 
* I used some file for the odbc-bridge-binary since I'm not running on windows and I couldn't find the binary.
* I had to add `--platform linux/amd64` for the command to run since the `clickhouse/integration-tests-runner` image is not multi-platform compatible and I'm running on M1 processor

```bash
➜ ./runner --binary ../../build/programs/clickhouse --base-configs-dir ../../build/programs/server/ --odbc-bridge-binary ../../build/programs/CTestTestfile.cmake 'test_storage_s3 -ss'
2023-12-18 16:21:50,037 [ 42429 ] INFO : ClickHouse root is not set. Will use /Users/shanielharrar/Documents/GitHub/ClickHouse (runner:42, check_args_and_update_paths)
2023-12-18 16:21:50,037 [ 42429 ] INFO : Cases dir is not set. Will use /Users/shanielharrar/Documents/GitHub/ClickHouse/tests/integration (runner:90, check_args_and_update_paths)
2023-12-18 16:21:50,037 [ 42429 ] INFO : utils dir is not set. Will use /Users/shanielharrar/Documents/GitHub/ClickHouse/utils (runner:101, check_args_and_update_paths)
2023-12-18 16:21:50,037 [ 42429 ] INFO : base_configs_dir: /Users/shanielharrar/Documents/GitHub/ClickHouse/build/programs/server, binary: /Users/shanielharrar/Documents/GitHub/ClickHouse/build/programs/clickhouse, cases_dir: /Users/shanielharrar/Documents/GitHub/ClickHouse/tests/integration  (runner:103, check_args_and_update_paths)
clickhouse_integration_tests_volume
Running pytest container as: 'docker run --net=host -it --rm --name clickhouse_integration_tests_tknz6q --platform linux/amd64 --privileged --dns-search='.' --volume=/Users/shanielharrar/Documents/GitHub/ClickHouse/build/programs/CTestTestfile.cmake:/clickhouse-odbc-bridge --volume=/Users/shanielharrar/Documents/GitHub/ClickHouse/build/programs/clickhouse:/clickhouse --volume=/Users/shanielharrar/Documents/GitHub/ClickHouse/build/programs/clickhouse-library-bridge:/clickhouse-library-bridge --volume=/Users/shanielharrar/Documents/GitHub/ClickHouse/build/programs/server:/clickhouse-config --volume=/Users/shanielharrar/Documents/GitHub/ClickHouse/tests/integration:/ClickHouse/tests/integration --volume=/Users/shanielharrar/Documents/GitHub/ClickHouse/utils/backupview:/ClickHouse/utils/backupview --volume=/Users/shanielharrar/Documents/GitHub/ClickHouse/utils/grpc-client/pb2:/ClickHouse/utils/grpc-client/pb2 --volume=/run:/run/host:ro --volume=clickhouse_integration_tests_volume:/var/lib/docker   -e DOCKER_CLIENT_TIMEOUT=300 -e COMPOSE_HTTP_TIMEOUT=600  -e PYTHONUNBUFFERED=1 -e PYTEST_ADDOPTS=" 'test_storage_s3 -ss'   -vvv" clickhouse/integration-tests-runner:latest '.
Can't start docker daemon, timeout exceeded.
Traceback (most recent call last):
  File "/Users/shanielharrar/Documents/GitHub/ClickHouse/tests/integration/./runner", line 461, in <module>
    subprocess.check_call(cmd, shell=True)
  File "/opt/homebrew/Cellar/python@3.11/3.11.6_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py", line 413, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'docker run --net=host -it --rm --name clickhouse_integration_tests_tknz6q --platform linux/amd64 --privileged --dns-search='.' --volume=/Users/shanielharrar/Documents/GitHub/ClickHouse/build/programs/CTestTestfile.cmake:/clickhouse-odbc-bridge --volume=/Users/shanielharrar/Documents/GitHub/ClickHouse/build/programs/clickhouse:/clickhouse --volume=/Users/shanielharrar/Documents/GitHub/ClickHouse/build/programs/clickhouse-library-bridge:/clickhouse-library-bridge --volume=/Users/shanielharrar/Documents/GitHub/ClickHouse/build/programs/server:/clickhouse-config --volume=/Users/shanielharrar/Documents/GitHub/ClickHouse/tests/integration:/ClickHouse/tests/integration --volume=/Users/shanielharrar/Documents/GitHub/ClickHouse/utils/backupview:/ClickHouse/utils/backupview --volume=/Users/shanielharrar/Documents/GitHub/ClickHouse/utils/grpc-client/pb2:/ClickHouse/utils/grpc-client/pb2 --volume=/run:/run/host:ro --volume=clickhouse_integration_tests_volume:/var/lib/docker   -e DOCKER_CLIENT_TIMEOUT=300 -e COMPOSE_HTTP_TIMEOUT=600  -e PYTHONUNBUFFERED=1 -e PYTEST_ADDOPTS=" 'test_storage_s3 -ss'   -vvv" clickhouse/integration-tests-runner:lates
```

**Comment by shanielh** [comment]
Ok @Avogar Seems like I forgot to change the number of arguments validation, so I just fixed it and I hope the CI is going to pass, I'll be happy if you can take a look at it today. Thank you.

**Review by Avogar — APPROVED** [review]

**Comment by UnamedRus** [comment]
Despite, session_token is being pushed to `auto credentials = Aws::Auth::AWSCredentials` struct,
It's not being provided for client itself : `    client = S3::ClientFactory::instance().create(...`
https://github.com/ClickHouse/ClickHouse/blob/63b445799136fa9fc4be362df35181064692fb48/src/Storages/StorageS3.cpp#L1434

However it should, there is separate overload of function for temporary creds with session_token:
https://github.com/ClickHouse/ClickHouse/blob/63b445799136fa9fc4be362df35181064692fb48/src/IO/S3/Client.cpp#L877C1-L877C34

**Comment by aser86** [comment]
Do a workaround exist to use session_token?

**Comment by UnamedRus** [comment]
You can pass them as enviromental variable, so aws sdk directly consume it from here

**Comment by aser86** [comment]
> You can pass them as enviromental variable, so aws sdk directly consume it from here

I passed the variables in this way:
export access_key_id=my_access_key_id
export secret_access_key=my_secret_access_key
export session_token=my_session_token

and copied the s3.xml to /opt/bitnami/clickhouse/etc/conf.d
but it doesn't work.

`
<clickhouse>
    <s3>
		<my_endpoint>
			<use_environment_credentials>true</use_environment_credentials>
		</my_endpoint>
    </s3>
</clickhouse>
`

I get this error message:

`
SQL-Fehler [499] [07000]: Code: 499. DB::Exception: Access Denied: while reading key: bucket_name/5014bb6a-263e-425c-98c0-97615db3fe3e.csv, from bucket: bucket_name: The data format cannot be detected by the contents of the files. You can specify the format manually: (in file/uri... (S3_ERROR) (version 24.2.1.2248 (official build))
, server ClickHouseNode [uri=http://localhost:8123/default, options={use_server_time_zone=false,use_time_zone=false}]@-1974819624
`

**Pull request merged**